[CIR][CIRGen][Builtin][X86] Masked compress Intrinsics

cs25resch11005-bhuvan opened the pull request:
Added masked compress builtin in CIR.
Note: This is my first PR to llvm. Looking forward to corrections

Bhuvan1527 committed:
[CIR][CIRGen][Builtin][X86] Masked compress Intrinsics

Added masked compress builtin in CIR.
Note: This is my first PR to llvm.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -454,7 +462,9 @@ mlir::Value CIRGenFunction::emitX86BuiltinExpr(unsigned builtinID,
   case X86::BI__builtin_ia32_compressqi128_mask:
   case X86::BI__builtin_ia32_compressqi256_mask:
-  case X86::BI__builtin_ia32_compressqi512_mask:
+  case X86::BI__builtin_ia32_compressqi512_mask:{
```
There is no need to add braces around this case.

HendrikHuebner commented on clang/test/CIR/CodeGen/X86/builtin_mask_compress.c:
```diff
@@ -0,0 +1,7 @@
+///home/cs25resch11005/myFiles/off_contrib/llvm-project/clang/lib/Headers/avx512vlvbmi2intrin.h
+
+__m128i test_mm_mask_compress(__m128i __S, __mmask8 __U, __m128i __D){
+
```
```suggestion
```

HendrikHuebner commented on clang/test/CIR/CodeGen/X86/builtin_mask_compress.c:
```diff
@@ -0,0 +1,7 @@
+#include <immintrin.h>
+///home/cs25resch11005/myFiles/off_contrib/llvm-project/clang/lib/Headers/avx512vlvbmi2intrin.h
```
```suggestion
```

HendrikHuebner commented on clang/test/CIR/CodeGen/X86/builtin_mask_compress.c:
```diff
@@ -0,0 +1,7 @@
+
+__m128i test_mm_mask_compress(__m128i __S, __mmask8 __U, __m128i __D){
+
+    return (__m128i)_mm_mask_compress_epi16(__S, __U, __D);
```
Please add assertions here. You can look at other tests for different builtins. Generally we do three checks:
1. Check that the correct clang IR assembly is generated (CIR checks)
2. Check that the correct LLVM IR is generated after lowering from CIR (LLVM checks)
3. Verify the LLVM IR generated with classing code gen (OGCG checks). This alerts us if something in classic codegen changes, since the tests will fail

HendrikHuebner commented:
You can autoformat the latest commit with this command:
```git diff -U0 --no-color --relative HEAD^ | <path-to-the-llvm-repo>/clang/tools/clang-format/clang-format-diff.py -p1 -i```

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -89,6 +89,14 @@ static mlir::Value getMaskVecValue(CIRGenFunction &cgf, const CallExpr *expr,
+  mlir::Value MaskValue = getMaskVecValue(cgf, expr, ops[2],  cast<cir::VectorType>(ResultTy).getSize());
+  llvm::SmallVector<mlir::Value, 4> op{ops[0], ops[1], MaskValue};
+  
+  return emitIntrinsicCallOp(cgf,expr, ID, ResultTy, op);
```
```suggestion
  return emitIntrinsicCallOp(cgf,expr, ID, ResultTy, { ops[0], ops[1], maskValue });
```
I don't think you need an explicit local to hold the array. Also, since you're passing the intrinsic name in as a string, the `isCompress` argument isn't needed.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
You need to add this here:

```
    cgm.errorNYI(expr->getSourceRange(),
                 std::string("unimplemented X86 builtin call: ") +
                     getContext().BuiltinInfo.getName(builtinID));
    return {};
```

Bhuvan1527 committed:
[CIR][CIRGen][Builtin][X86] Masked compress Intrinsics

This pr is related to the issue #167765
Added the support Masked compress builtin in CIR codeGen

cs25resch11005-bhuvan commented on clang/lib/CIR/CodeGen/CIRGenBuiltinX86.cpp:
```diff
@@ -89,6 +89,14 @@ static mlir::Value getMaskVecValue(CIRGenFunction &cgf, const CallExpr *expr,
   }
   return maskVec;
 }
+static mlir::Value emitX86CompressExpand(CIRGenFunction &cgf, const CallExpr *expr,ArrayRef<mlir::Value> ops, bool IsCompress, const std::string &ID){
```
Addressed in the latest commit

cs25resch11005-bhuvan commented on clang/test/CIR/CodeGen/X86/builtin_mask_compress.c:
```diff
@@ -0,0 +1,7 @@
+#include <immintrin.h>
```
@andykaylor could you please suggest a way to add these lines? Do we have any
 automated scripts which can help.

andykaylor commented on clang/test/CIR/CodeGenBuiltins/X86/avx512vl-builtins.c:
```diff
@@ -199,3 +199,36 @@ __m256i test_mm256_mask_i32gather_epi32(__m256i __v1_old, __mmask8 __mask, __m25
+
+  // OGCG-LABEL: test_mm_mask_expand_pd
+  // OGCG: %[[BC2:.*]] = bitcast i8 %{{.*}} to <8 x i1>
+  // OGCG: %[[SHUF2:.*]] = shufflevector <8 x i1> %[[BC2]], <8 x i1> %[[BC2]], <2 x i32> <i32 0, i32 1>
```
```suggestion
  // OGCG: %[[BC:.*]] = bitcast i8 %{{.*}} to <8 x i1>
  // OGCG: %[[SHUF:.*]] = shufflevector <8 x i1> %[[BC2]], <8 x i1> %[[BC2]], <2 x i32> <i32 0, i32 1>
```
There's no need to unique these capture identifiers. This will be used in a separate FileCheck run, and even if they weren't it's OK to overwrite variables in separate tests.

andykaylor reviewed (commented):
The checks look good. I have just a few comments.

Bhuvan1527 committed:
[CIR][CIRGen][Builtin][X86] Masked compress Intrinsics

Added masked compress builtin in CIR.Fixed the clang-format issues

andykaylor reviewed (approved):
lgtm

andykaylor reviewed (commented):
It looks like the test needs an update. Can you also fix the code formatting to conform with clang-format's suggestion. Thanks!

Bhuvan1527 committed:
[CIR][CIRGen][Builtin][X86] Masked compress Intrinsics

Added masked compress builtin in CIR.
